File: Languages/Java/HashGen.java
import java.io.*;
import java.nio.file.*;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.stream.Collectors;

public class HashGen {
    
    public static void main(String[] args) {
        System.out.println("Hello, World!");
        hashFilesInDirectory("../hashfiles");
    }
    
    private static class FileHashResult {
        String filename;
        String filepath;
        long size;
        String hash;
        String status;
        String error;
        
        FileHashResult(String filename, String filepath, long size) {
            this.filename = filename;
            this.filepath = filepath;
            this.size = size;
            this.status = "SUCCESS";
        }
    }
    
    private static class HashResults {
        int totalFiles;
        int successfulHashes;
        int failedHashes;
        double processingTimeSeconds;
        double averageTimePerFileMs;
        String timestamp;
        String directory;
        List<FileHashResult> results;
        
        HashResults(String directory) {
            this.directory = directory;
            this.results = new ArrayList<>();
        }
    }
    
    private static String calculateFileHash(String filepath) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] fileBytes = Files.readAllBytes(Paths.get(filepath));
            byte[] hashBytes = digest.digest(fileBytes);
            
            StringBuilder hexString = new StringBuilder();
            for (byte b : hashBytes) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) {
                    hexString.append('0');
                }
                hexString.append(hex);
            }
            
            return hexString.toString();
        } catch (NoSuchAlgorithmException | IOException e) {
            return "";
        }
    }
    
    private static long getFileSize(String filepath) {
        try {
            return Files.size(Paths.get(filepath));
        } catch (IOException e) {
            return 0;
        }
    }
    
    private static List<String> getTxtFiles(String directory) {
        try {
            Path dirPath = Paths.get(directory);
            if (!Files.exists(dirPath) || !Files.isDirectory(dirPath)) {
                System.out.println("Directory not found: " + directory);
                return Collections.emptyList();
            }
            
            return Files.list(dirPath)
                    .filter(path -> Files.isRegularFile(path) && path.toString().endsWith(".txt"))
                    .map(Path::toString)
                    .sorted()
                    .collect(Collectors.toList());
        } catch (IOException e) {
            System.out.println("Error reading directory: " + e.getMessage());
            return Collections.emptyList();
        }
    }
    
    private static String getTimestamp() {
        return new java.util.Date().toString();
    }
    
    private static String escapeJson(String str) {
        if (str == null) return "";
        return str.replace("\\", "\\\\")
                  .replace("\"", "\\\"")
                  .replace("\n", "\\n")
                  .replace("\r", "\\r")
                  .replace("\t", "\\t");
    }
    
    private static void hashFilesInDirectory(String directory) {
        List<String> files = getTxtFiles(directory);
        
        if (files.isEmpty()) {
            System.out.println("No .txt files found in directory: " + directory);
            return;
        }
        
        System.out.println("Found " + files.size() + " files to hash");
        
        HashResults results = new HashResults(directory);
        results.totalFiles = files.size();
        
        long startTime = System.currentTimeMillis();
        
        for (int i = 0; i < files.size(); i++) {
            String filepath = files.get(i);
            FileHashResult result = new FileHashResult(
                Paths.get(filepath).getFileName().toString(),
                filepath,
                getFileSize(filepath)
            );
            
            String hash = calculateFileHash(filepath);
            if (!hash.isEmpty()) {
                result.hash = hash;
                result.status = "SUCCESS";
                results.successfulHashes++;
            } else {
                result.hash = "";
                result.status = "FAILED";
                result.error = "Hash calculation failed";
                results.failedHashes++;
            }
            
            results.results.add(result);
            
            // Progress indicator
            if ((i + 1) % 100 == 0) {
                System.out.println("Processed " + (i + 1) + "/" + files.size() + " files...");
            }
        }
        
        long endTime = System.currentTimeMillis();
        results.processingTimeSeconds = (endTime - startTime) / 1000.0;
        results.averageTimePerFileMs = (results.processingTimeSeconds / files.size()) * 1000.0;
        results.timestamp = getTimestamp();
        
        // Create JSON output
        StringBuilder jsonBuilder = new StringBuilder();
        jsonBuilder.append("{\n");
        jsonBuilder.append("  \"total_files\": ").append(results.totalFiles).append(",\n");
        jsonBuilder.append("  \"successful_hashes\": ").append(results.successfulHashes).append(",\n");
        jsonBuilder.append("  \"failed_hashes\": ").append(results.failedHashes).append(",\n");
        jsonBuilder.append("  \"algorithm\": \"sha256\",\n");
        jsonBuilder.append("  \"processing_time_seconds\": ").append(results.processingTimeSeconds).append(",\n");
        jsonBuilder.append("  \"average_time_per_file_ms\": ").append(results.averageTimePerFileMs).append(",\n");
        jsonBuilder.append("  \"timestamp\": \"").append(results.timestamp).append("\",\n");
        jsonBuilder.append("  \"directory\": \"").append(results.directory).append("\",\n");
        jsonBuilder.append("  \"results\": [\n");
        
        for (int i = 0; i < results.results.size(); i++) {
            FileHashResult result = results.results.get(i);
            jsonBuilder.append("    {\n");
            jsonBuilder.append("      \"filename\": \"").append(escapeJson(result.filename)).append("\",\n");
            jsonBuilder.append("      \"filepath\": \"").append(escapeJson(result.filepath)).append("\",\n");
            jsonBuilder.append("      \"size\": ").append(result.size).append(",\n");
            jsonBuilder.append("      \"algorithm\": \"sha256\",\n");
            jsonBuilder.append("      \"hash\": \"").append(result.hash).append("\",\n");
            jsonBuilder.append("      \"status\": \"").append(result.status).append("\"");
            
            if (result.error != null) {
                jsonBuilder.append(",\n      \"error\": \"").append(escapeJson(result.error)).append("\"");
            }
            
            jsonBuilder.append("\n    }");
            
            if (i < results.results.size() - 1) {
                jsonBuilder.append(",");
            }
            jsonBuilder.append("\n");
        }
        
        jsonBuilder.append("  ]\n");
        jsonBuilder.append("}");
        
        // Save to file
        try (FileWriter file = new FileWriter("hash_results.json")) {
            file.write(jsonBuilder.toString());
            System.out.println("Results saved to: hash_results.json");
        } catch (IOException e) {
            System.out.println("Error saving results: " + e.getMessage());
        }
        
        System.out.println("\nHashing completed!");
        System.out.println("Total files: " + results.totalFiles);
        System.out.println("Successful: " + results.successfulHashes);
        System.out.println("Failed: " + results.failedHashes);
        System.out.printf("Processing time: %.3f seconds\n", results.processingTimeSeconds);
        System.out.printf("Average time per file: %.2f ms\n", results.averageTimePerFileMs);
    }
}

File: Languages/Javascript/hashgen.js
// Hello, World!
console.log("Hello, World!");

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

class FileHashResult {
    constructor(filename, filepath, size) {
        this.filename = filename;
        this.filepath = filepath;
        this.size = size;
        this.status = "SUCCESS";
    }
}

class HashResults {
    constructor(directory) {
        this.directory = directory;
        this.results = [];
    }
}

function calculateFileHash(filepath) {
    try {
        const fileBuffer = fs.readFileSync(filepath);
        const hashSum = crypto.createHash('sha256');
        hashSum.update(fileBuffer);
        return hashSum.digest('hex');
    } catch (error) {
        return '';
    }
}

function getFileSize(filepath) {
    try {
        const stats = fs.statSync(filepath);
        return stats.size;
    } catch (error) {
        return 0;
    }
}

function getTxtFiles(directory) {
    try {
        if (!fs.existsSync(directory) || !fs.statSync(directory).isDirectory()) {
            console.log(`Directory not found: ${directory}`);
            return [];
        }
        
        const files = fs.readdirSync(directory)
            .filter(file => file.endsWith('.txt'))
            .map(file => path.join(directory, file))
            .sort();
        
        return files;
    } catch (error) {
        console.log(`Error reading directory: ${error.message}`);
        return [];
    }
}

function getTimestamp() {
    return new Date().toISOString();
}

function hashFilesInDirectory(directory) {
    const files = getTxtFiles(directory);
    
    if (files.length === 0) {
        console.log(`No .txt files found in directory: ${directory}`);
        return;
    }
    
    console.log(`Found ${files.length} files to hash`);
    
    const results = new HashResults(directory);
    results.totalFiles = files.length;
    
    const startTime = Date.now();
    
    for (let i = 0; i < files.length; i++) {
        const filepath = files[i];
        const result = new FileHashResult(
            path.basename(filepath),
            filepath,
            getFileSize(filepath)
        );
        
        const hash = calculateFileHash(filepath);
        if (hash) {
            result.hash = hash;
            result.status = "SUCCESS";
            results.successfulHashes = (results.successfulHashes || 0) + 1;
        } else {
            result.hash = "";
            result.status = "FAILED";
            result.error = "Hash calculation failed";
            results.failedHashes = (results.failedHashes || 0) + 1;
        }
        
        results.results.push(result);
        
        // Progress indicator
        if ((i + 1) % 100 === 0) {
            console.log(`Processed ${i + 1}/${files.length} files...`);
        }
    }
    
    const endTime = Date.now();
    results.processingTimeSeconds = (endTime - startTime) / 1000;
    results.averageTimePerFileMs = (results.processingTimeSeconds / files.length) * 1000;
    results.timestamp = getTimestamp();
    
    // Create JSON output
    const jsonResults = {
        total_files: results.totalFiles,
        successful_hashes: results.successfulHashes || 0,
        failed_hashes: results.failedHashes || 0,
        algorithm: "sha256",
        processing_time_seconds: results.processingTimeSeconds,
        average_time_per_file_ms: results.averageTimePerFileMs,
        timestamp: results.timestamp,
        directory: results.directory,
        results: results.results.map(result => {
            const jsonResult = {
                filename: result.filename,
                filepath: result.filepath,
                size: result.size,
                algorithm: "sha256",
                hash: result.hash,
                status: result.status
            };
            
            if (result.error) {
                jsonResult.error = result.error;
            }
            
            return jsonResult;
        })
    };
    
    // Save to file
    try {
        fs.writeFileSync('hash_results.json', JSON.stringify(jsonResults, null, 2));
        console.log('Results saved to: hash_results.json');
    } catch (error) {
        console.log(`Error saving results: ${error.message}`);
    }
    
    console.log('\nHashing completed!');
    console.log(`Total files: ${results.totalFiles}`);
    console.log(`Successful: ${results.successfulHashes || 0}`);
    console.log(`Failed: ${results.failedHashes || 0}`);
    console.log(`Processing time: ${results.processingTimeSeconds.toFixed(3)} seconds`);
    console.log(`Average time per file: ${results.averageTimePerFileMs.toFixed(2)} ms`);
}

hashFilesInDirectory('../hashfiles');
